river/LayoutManager.zig

```diff
@@ -0,0 +1,378 @@
+        if (output.layout_demand) |demand| {
+            util.gpa.destroy(demand);
+            output.layout_demand = null;
+        }
+
+        // Start a transaction, so things that happened while
+        // this LayoutDemand was timing out are applied correctly.
+        // Remember: An active layout demand blocks all transactions.
+        // on the affected output.
+        output.root.startTransaction();
+
+        return 0;
+    }
+
+    /// Push a set of proposed view dimensions and position to the list
+    pub fn pushViewDimensions(self: *LayoutDemand, x: i32, y: i32, width: u32, height: u32) !void {
+        const node = try util.gpa.create(std.TailQueue(Box).Node);
+
+        // Here we apply the offset to align the coords with the origin of the
+        // usable area and shrink the dimensions to accomodate the border size.
+        node.data = .{
+            .x = x + self.output.usable_box.x + @intCast(i32, self.output.root.server.config.border_width),
+            .y = y + self.output.usable_box.y + @intCast(i32, self.output.root.server.config.border_width),
+            .width = width - 2 * self.output.root.server.config.border_width,
+            .height = height - 2 * self.output.root.server.config.border_width,
+        };
+
+        self.view_dimensions.append(node);
+    }
+
+    /// Get a set of proposed view dimensions and position from the list, starting at the top
+    pub fn popViewDimensions(self: *LayoutDemand) ?Box {
+        if (self.view_dimensions.popFirst()) |node| {
+            const box: Box = node.data;
+            util.gpa.destroy(node);
+            return box;
+        }
+
+        return null;
+    }
+};
+
+/// This object encapsulates a zriver.LayoutV1 object bound by a client
+pub const Layout = struct {
+    layout: *zriver.LayoutV1 = undefined,
+    namespace: ?[]const u8 = null,
+    output: *Output = undefined,
+
+    pub fn init(self: *Layout, namespace: [*:0]const u8, output: *Output, client: *wl.Client, version: u32, id: u32) !void {
+        self.* = .{
+            .output = output,
+            .namespace = try util.gpa.dupe(u8, std.mem.span(namespace)),
+        };
+
+        self.layout = try zriver.LayoutV1.create(client, version, id);
+        self.layout.setHandler(*Layout, handleRequest, null, self);
+
+        // Namespaces must be unique, so if any other layout on this output
+        // already has the requested namespace it can not be chosen by this
+        // one. In that case we send the namespace_in_use event to the client
+        // and free the namespace buffer, marking the Layout object as invalid.
+        var it = output.layouts.first;
+        while (it) |layout_node| : (it = layout_node.next) {
+            if (std.mem.eql(u8, self.namespace.?, layout_node.data.namespace.?)) {
+                self.layout.sendNamespaceInUse();
+                util.gpa.free(self.namespace.?);
+                self.namespace = null;
+                return;
+            }
+        }
```

Comment:
Not sure if you saw my comment on IRC, but I think enforcing globally unique namespaces where the same client may register the same namespace on multiple outputs would be the most intuitive way to handle this. This would mean iterating over all layout objects on all outputs and only allowing a duplicate if `layout.getClient()` returns the same `wl.Client` pointer.